src/Diagnostics.RuntimeHost/Controllers/Configuration/KustoClusterMappingsController.cs

```diff
@@ -0,0 +1,85 @@
+﻿// <copyright file="KustoClusterMappingsController.cs" company="Microsoft Corporation">
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Diagnostics.RuntimeHost.Models;
+using Diagnostics.RuntimeHost.Services.CacheService;
+using Diagnostics.RuntimeHost.Services.SourceWatcher;
+using Diagnostics.RuntimeHost.Utilities;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Newtonsoft.Json;
+
+namespace Diagnostics.RuntimeHost.Controllers.Configuration
+{
+    [Authorize]
+    [Produces("application/json")]
+    [Route(UriElements.ArmResource + "/" + UriElements.KustoClusterMappings)]
+    public class KustoClusterMappingsController : Controller
+    {
+        private ISourceWatcherService _sourceWatcherService;
+        private IKustoMappingsCacheService _kustoMappingsCache;
+
+        public KustoClusterMappingsController(ISourceWatcherService sourceWatcherService, IKustoMappingsCacheService kustoMappingsCache)
+        {
+            _sourceWatcherService = sourceWatcherService;
+            _kustoMappingsCache = kustoMappingsCache;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> AddOrUpdateMapping(string provider, [FromBody]List<Dictionary<string, string>> kustoMappings)
+        {
+            var cacheId = GetGitHubId(provider);
+            await _sourceWatcherService.Watcher.WaitForFirstCompletion();
+
+            //TODO the equals condition always fail due to the strings in the data structures are never compared
+            if (_kustoMappingsCache.TryGetValue(cacheId, out List<Dictionary<string, string>> cacheValue) && cacheValue.Equals(kustoMappings))
+            {
+                return Ok();
+            }
+
+            var gitHubPackage = new GithubPackage(cacheId, "kustoClusterMappings", "json", JsonConvert.SerializeObject(kustoMappings));
+
+            try
+            {
+                await _sourceWatcherService.Watcher.CreateOrUpdatePackage(gitHubPackage);
```

Comment:
Should we start the try in the beginning itself because we have `_sourceWatcherService` call before this

Reply:
Good idea